app/src/ui/dropdown-select-button.tsx

```diff
@@ -0,0 +1,203 @@
+}
+
+export class DropdownSelectButton extends React.Component<
+  IDropdownSelectButtonProps,
+  IDropdownSelectButtonState
+> {
+  private invokeButtonRef: HTMLButtonElement | null = null
+  private optionsContainerRef: HTMLDivElement | null = null
+
+  public constructor(props: IDropdownSelectButtonProps) {
+    super(props)
+
+    this.state = {
+      showButtonOptions: false,
+      selectedOption: this.getSelectedOption(props.selectedValue),
+    }
+  }
+
+  public componentDidUpdate() {
+    if (this.invokeButtonRef === null || this.optionsContainerRef === null) {
+      return
+    }
+
+    const windowHeight = window.innerHeight
+    const bottomOfButton = this.invokeButtonRef.getBoundingClientRect().bottom
+    const invokeButtonHeight = this.invokeButtonRef.getBoundingClientRect()
+      .height
+    // 15 pixels is just to give some padding room below it
+    const calcMaxHeight = windowHeight - bottomOfButton - 15
+    const heightOfOptions = this.optionsContainerRef.clientHeight
+    const optionsPositionBottom =
+      calcMaxHeight < heightOfOptions ? `${invokeButtonHeight}px` : undefined
+    if (optionsPositionBottom !== this.state.optionsPositionBottom) {
+      this.setState({ optionsPositionBottom })
+    }
+  }
+
+  private getSelectedOption(
+    selectedValue: string | undefined
+  ): IDropdownSelectButtonOption | null {
+    const { options } = this.props
+    if (options.length === 0) {
+      return null
+    }
+
+    const selectedOption = options.find(o => o.value === selectedValue)
+    if (selectedOption === undefined) {
+      return options[0]
+    }
+    return selectedOption
+  }
+
+  private onSelectionChange = (selectedOption: IDropdownSelectButtonOption) => {
+    return (_event: React.MouseEvent<HTMLLIElement, MouseEvent>) => {
+      this.setState({ selectedOption, showButtonOptions: false })
+
+      const { onSelectChange } = this.props
+      if (onSelectChange !== undefined) {
+        onSelectChange(selectedOption)
+      }
+    }
+  }
+
+  private openSplitButtonDropdown = () => {
+    this.setState({ showButtonOptions: !this.state.showButtonOptions })
+  }
+
+  private onInvokeButtonRef = (buttonRef: HTMLButtonElement | null) => {
+    this.invokeButtonRef = buttonRef
+  }
+
+  private onOptionsContainerRef = (ref: HTMLDivElement | null) => {
+    this.optionsContainerRef = ref
+  }
+
+  private renderSelectedIcon(option: IDropdownSelectButtonOption) {
+    const { selectedOption } = this.state
+    if (selectedOption === null || option.value !== selectedOption.value) {
+      return
+    }
+
+    return (
+      <Octicon
+        className="selected-option-indicator"
+        symbol={OcticonSymbol.check}
+      />
+    )
+  }
+
+  private renderSplitButtonOptions() {
+    if (!this.state.showButtonOptions) {
+      return
+    }
+
+    const { options } = this.props
+    const { optionsPositionBottom: bottom } = this.state
+    const openClass = bottom !== undefined ? 'open-top' : 'open-bottom'
```

Comment:
I think `optionsPositionBottom` and `bottom` are a bit misleading in this case, because I was expecting this condition to be the other way round 😂  
```suggestion
    const openClass = bottom !== undefined ? 'open-bottom' : 'open-top'
```

Reply:
Yeah. Definitely confusing.. but the definition of a bottom position style means that we are setting the 'bottom' style property of the options so that it will open to the top. I may tweak this logic (later) to store a boolean and calculate the bottom at the time usage to make reading it clearer.